test/unit_test/runtime/magnitude_test.cpp

```diff
@@ -0,0 +1,403 @@
+  {
+    CHECK(strictly_increasing(std::make_tuple(3)));
+    CHECK(strictly_increasing(std::make_tuple(15.42)));
+    CHECK(strictly_increasing(std::make_tuple('c')));
+  }
+
+  SECTION ("Multi-element tuples compare correctly")
+  {
+    CHECK(strictly_increasing(std::make_tuple(3, 3.14)));
+    CHECK(!strictly_increasing(std::make_tuple(3, 3.0)));
+    CHECK(!strictly_increasing(std::make_tuple(4, 3.0)));
+  }
+}
+
+TEST_CASE("make_ratio performs prime factorization correctly")
+{
+  SECTION("Performs prime factorization when denominator is 1")
+  {
+    CHECK(std::is_same_v<decltype(make_ratio<1>()), magnitude<>>);
+    CHECK(std::is_same_v<decltype(make_ratio<2>()), magnitude<int_base_power<2>>>);
+    CHECK(std::is_same_v<decltype(make_ratio<3>()), magnitude<int_base_power<3>>>);
+    CHECK(std::is_same_v<decltype(make_ratio<4>()), magnitude<int_base_power<2, 2>>>);
+
+    CHECK(std::is_same_v<
+        decltype(make_ratio<792>()),
+        magnitude<int_base_power<2, 3>, int_base_power<3, 2>, int_base_power<11>>>);
+  }
+
+  SECTION("Reduces fractions to lowest terms")
+  {
+    CHECK(std::is_same_v<decltype(make_ratio<8, 8>()), magnitude<>>);
+    CHECK(std::is_same_v<
+        decltype(make_ratio<50, 80>()),
+        magnitude<int_base_power<2, -3>, int_base_power<5>>>);
+  }
+}
+
+TEST_CASE("make_magnitude handles arbitrary bases")
+{
+  SECTION("Equivalent to std::integral_constant for integer bases")
+  {
+    CHECK(make_base_power<prime_base<2>>() == make_ratio<2>());
+    CHECK(make_base_power<prime_base<7>>() == make_ratio<7>());
+  }
+
+  SECTION("Handles non-integer bases")
+  {
+    CHECK(make_base_power<pi>() == magnitude<base_power<pi>>{});
+    CHECK(make_base_power<pi, -3>() == magnitude<base_power<pi, ratio{-3}>>{});
+    CHECK(make_base_power<pi, -3, 7>() == magnitude<base_power<pi, ratio{-3, 7}>>{});
+  }
+}
+
+TEST_CASE("Equality works for magnitudes")
+{
+  SECTION("Equivalent ratios are equal")
+  {
+    CHECK(make_ratio<1>() == make_ratio<1>());
+    CHECK(make_ratio<3>() == make_ratio<3>());
+    CHECK(make_ratio<3, 4>() == make_ratio<9, 12>());
+  }
+
+  SECTION("Different ratios are unequal")
+  {
+    CHECK(make_ratio<3>() != make_ratio<5>());
+    CHECK(make_ratio<3>() != make_ratio<3, 2>());
+  }
+
+  SECTION("Supports constexpr")
+  {
+    constexpr auto eq = make_ratio<4, 5>() == make_ratio<4, 3>();
+    CHECK(!eq);
+  }
+}
+
+TEST_CASE("Multiplication works for magnitudes")
+{
+  SECTION("Reciprocals reduce to null magnitude")
+  {
+    CHECK(make_ratio<3, 4>() * make_ratio<4, 3>() == make_ratio<1>());
+  }
+
+  SECTION("Products work as expected")
+  {
+    CHECK(make_ratio<4, 5>() * make_ratio<4, 3>() == make_ratio<16, 15>());
+  }
+
+  SECTION("Products handle pi correctly")
+  {
+    CHECK(
+        make_base_power<pi>() * make_ratio<2, 3>() * make_base_power<pi, -1, 2>() ==
+        magnitude<int_base_power<2>, int_base_power<3, -1>, base_power<pi, ratio{1, 2}>>{});
```

Comment:
Does it mean that the `Den` part of `ratio` is only needed for floating-point bases (for integrals you can always extract num and den to separate magnitudes)? Also, it seems that `exp` of `units::ratio` is never used here. Maybe `units::ratio` is not the best to implement this (for sure it makes the type name longer in the compilation errors).

Reply:
Also, the fact of having `exp` in `units::ratio` complicates many operations which could be faster to compile (or more accurate) if it was not there. If we can simplify things there it would be great.

Reply:
> `Den` is not 1, and there's no other way to represent this, right?

Right, I forgot about `sqrt`...

Reply:
I would love to be able to get rid of `exp` from `ratio` if only possible.

Reply:
I don't know if it will work, because I haven't looked at all the use cases in the codebase.  But I plan to _try_ to get rid of it, and I expect to succeed!